### PR TITLE
Implement Quality of Live features for instrumentation

### LINKED ISSUE #314: Remove unit-testing-tracing-auto.ipf
This file is a bit ugly as it is checked in, but is also modified by us.

Can't we replace the functionality of `GetTracedProcedureNames()` with a wave stored in the TUF XOP?

### LINKED ISSUE #315: Add utility function to restore backups for traced procedures

### PR DESCRIPTION
**Pull request by Garados007**
This consists of the following changes:

- **Add restore of tracing backup function**

    Tracing does automatically a backup for all instrumented files. A new function is added to restore these backups and reset the instrumentation.
- **Remove unit-testing-tracing-auto.ipf in favor of TUFXOP**

    unit-testing-tracing-auto.ipf stored the list of instrumented procedure files as a readonly threadsafe free wave. This file is autogenerated and stored alongside UTF itself. This has some issues as it is difficult to manage inside a versioning system and it provokes problems when different experiments are tested at the same time.

    The data is stored now in a global wave inside memory using TUFXOP and accessible whenever it's needed. This data is no longer written to disk.

Close #314
Close #315

### CODE REVIEW

**Comment by Garados007** [comment]
@t-b @MichaelHuth Ready for review.

File: procedures/unit-testing-tracing.ipf [hunk 1]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+		path = GetProcedurePath(wvProcs[i])
+		backupPath = path + ".backup"
+
+		GetFileFolderInfo/Q/Z=1 backupPath
+		if(V_flag)
+			sprintf msg, "Backup file not found: %s", backupPath
+			UTF_Reporting#UTF_PrintStatusMessage(msg)
+			continue
+		else
+			sprintf msg, "Backup restored for %s", path
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:988]: Thats too early here. Should be after `MoveFile`

File: procedures/unit-testing-tracing.ipf [hunk 2]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+		sprintf msg, "Procedure window %s not found.", procWin
+		UTF_Reporting#ReportErrorAndAbort(msg)
+	endif
+
+	macrosList = MacroList("*", ";", "KIND:7,WIN:" + procWin)
+	funcList = FunctionList("*", ";", "KIND:18,WIN:" + procWin)
+
+	WAVE/T wMacroList = ListToTextWave(macrosList, ";")
+	size = DimSize(wMacroList, UTF_ROW)
+	for(i = 0; i < size; i += 1)
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:1019]: `for(macroName : wMacroList)`

File: procedures/unit-testing-tracing.ipf [hunk 3]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+		path = MacroPath(wMacroList[i])
+		if(!UTF_Utils#isEmpty(path))
+			return path
+		endif
+	endfor
+
+	WAVE/T wFuncList = ListToTextWave(funcList, ";")
+	size = DimSize(wFuncList, UTF_ROW)
+	for(i = 0; i < size; i += 1)
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:1028]: `for(funcName : wFuncList)`

File: procedures/unit-testing-tracing.ipf [hunk 4]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+		name = UTF_Basics#getFullFunctionName(err, wFuncList[i], procWin)
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:1029]: If you name the variable here `fullFuncName` it is better described.

File: procedures/unit-testing-tracing.ipf [hunk 5]
```diff
@@ -186,6 +186,8 @@ static Function SetupTraceProcedures(string procWinList, string traceOptions)
 
 	iProcList = ""
 
+	printf "procWinList: %s\n", procWinList
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:195]: Is this some debugging residue?

File: procedures/unit-testing-tracing.ipf [hunk 6]
```diff
@@ -280,26 +284,33 @@ static Function WriteProcList(string procWinList)
...
 
-	newCode = "Make/FREE/T wt = {\\\r"
-	for(i = 0; i < numProcs - 1; i += 1)
-		newCode += "\"" + StringFromList(i, procWinList) + "\",\\\r"
-	endfor
-	newCode += "\"" + StringFromList(numProcs - 1, procWinList) + "\"\\\r"
-	newCode += "}\r"
-	newCode += "return wt\r"
-	WAVE/T wNewCode = ListToTextWave(newCode, "\r")
+	if(numProcs)
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf]: If you fix bugs in old code put that in a separate commit.

File: procedures/unit-testing-tracing.ipf [hunk 7]
```diff
@@ -280,26 +284,33 @@ static Function WriteProcList(string procWinList)
...
+		newCode = "Make/FREE/T wt = {\\\n"
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf]: What is the change of all line endings fixing here?

File: procedures/unit-testing-tracing.ipf [hunk 8]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+			UTF_Reporting#UTF_PrintStatusMessage(msg)
+			continue
+		else
+			sprintf msg, "Backup restored for %s", path
+			UTF_Reporting#UTF_PrintStatusMessage(msg)
+		endif
+
+		MoveFile/O/Z=1 backupPath as path
+		if(V_flag)
+			sprintf msg, "Cannot move %s to %s (error: %d)", S_fileName, S_path, V_flag
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:994]: You can use `GetErrMessage` to get a text for the error code of internal functions

File: procedures/unit-testing-tracing.ipf [hunk 9]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+	WAVE/T wMacroList = ListToTextWave(macrosList, ";")
+	size = DimSize(wMacroList, UTF_ROW)
+	for(i = 0; i < size; i += 1)
+		path = MacroPath(wMacroList[i])
+		if(!UTF_Utils#isEmpty(path))
+			return path
+		endif
+	endfor
+
+	WAVE/T wFuncList = ListToTextWave(funcList, ";")
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:1026]: I would do function evaluation first, as for the general case it is much more likely that a procedure file has functions than macros.

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
Can you give some reasoning why you chose to resolve the procedure path again for all procwins to reach the backup files?

You could have also saved the paths in SetupTraceProcedures where the backups are created in a wave. Would that cause some issue?

**Comment by Garados007** [comment]
> Can you give some reasoning why you chose to resolve the procedure path again for all procwins to reach the backup files?

It's about the ordering of the commits. In the first commit I need the files paths but I only have the procedure names stored in `auto.ipf`. In the second commit I remove the `auto.ipf` file at all.

I think I will switch the order of the commits and prevent this problem.

File: procedures/unit-testing-tracing.ipf (hunk 6 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: After reordering of commits this is outdated.

File: procedures/unit-testing-tracing.ipf (hunk 7 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: The original idea was that this function produce the same empty file as it is currently checked in. The current line ending prevented this so I have fixed it. After reordering the commit this change does no longer exists.

File: procedures/unit-testing-tracing.ipf (hunk 5 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: Yes

File: procedures/unit-testing-tracing.ipf [hunk 10]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+
+	if(size == 1 && !strlen(wvProcs[0]))
+		UTF_Reporting#UTF_PrintStatusMessage("Nothing to restore")
+		return NaN
+	endif
+
+	sprintf msg, "%d procedure files to restore", size
+	UTF_Reporting#UTF_PrintStatusMessage(msg)
+
+	for(i = 0; i < size; i += 1)
```
[the remark this answers is not in the document]
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf:1020]: I didn't know this was a thing in IP9 but I don't think this will work with the newest changes (wvProcs could now be larger than size).

File: procedures/unit-testing-tracing.ipf [hunk 11]
```diff
@@ -178,6 +172,18 @@ static Function/S PreCheckProcedures(string procWinList)
 	return outList
 End
 
+/// @brief Returns the traced procedure names. Do not modify these!
+threadsafe static Function/WAVE GetTracedProcedureNames()
+	TUFXOP_GetStorage/N="IUTF_Traced_Procedures" wvStorage
+	if(V_flag)
+		Make/FREE=1/T/N=0 empty
```
[the remark this answers is not in the document]
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf:193]: done

File: procedures/unit-testing-tracing.ipf [hunk 12]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
+	endif
+
+	sprintf msg, "%d procedure files to restore", size
+	UTF_Reporting#UTF_PrintStatusMessage(msg)
+
+	for(i = 0; i < size; i += 1)
+		path = GetProcedurePath(wvProcs[i])
+		backupPath = path + ".backup"
+
+		GetFileFolderInfo/Q/Z=1 backupPath
```
[the remark this answers is not in the document]
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: done

File: procedures/unit-testing-tracing.ipf (hunk 1 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: my bad

File: procedures/unit-testing-tracing.ipf (hunk 2 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: After reordering the commit this function does no longer exists.

File: procedures/unit-testing-tracing.ipf (hunk 3 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: After reordering the commit this function does no longer exists.

File: procedures/unit-testing-tracing.ipf (hunk 4 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: I applied your suggestion storing the file path with the procedure names. There was no longer a need for this function and I have removed it.

File: procedures/unit-testing-tracing.ipf (hunk 8 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: Thanks

File: procedures/unit-testing-tracing.ipf (hunk 9 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: I applied your suggestion storing the file path with the procedure names. There was no longer a need for this function and I have removed it.

File: procedures/unit-testing-tracing.ipf [hunk 13]
```diff
@@ -988,4 +999,88 @@ static Function AnalyzeTracingResult()
...
 End
 
+Function IUTF_RestoreTracing()
+	variable i, size
+	string path, backupPath, msg
+
+	WAVE/T wvProcs = GetTracedProcedureNames()
+	size = DimSize(wvProcs, UTF_ROW)
+
+	if(size == 1 && !strlen(wvProcs[0]))
```
[the remark this answers is not in the document]
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: fixed it

**Comment by Garados007** [comment]
Fixed all the stuff. Please review again. :D

File: procedures/unit-testing-tracing.ipf [hunk 14]
```diff
@@ -172,9 +172,22 @@ static Function/S PreCheckProcedures(string procWinList)
 	return outList
 End
 
+/// @brief Returns the traced procedures. Do not modify these!
+threadsafe static Function/WAVE GetTracedProcedures()
+	TUFXOP_GetStorage/N="IUTF_Traced_Procedures" wvStorage
+	if(V_flag)
+		UTF_Reporting#UTF_PrintStatusMessage("Error: Cannot get IUTF_Traced_ProcedureNames storage for traced procedure names")
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:179]: The error message here is a copy of the one from GetTracedProcedureNames and should probably be different.
How about naming this GetTracedProcedurePaths to be more distinct from the other global.

While the procedure names must be available for all threads the paths don't have to be. It is easier to debug if stored in the IUTF Package folder.

File: procedures/unit-testing-tracing.ipf [hunk 15]
```diff
@@ -254,24 +273,49 @@ static Function SetupTraceProcedures(string procWinList, string traceOptions)
...
+		sprintf msg, "Cannot open shared wave for procedures (error: %d)", V_flag
+		UTF_Reporting#ReportErrorAndAbort(msg)
+	endif
+	wvStorage[0] = procTextGrid
+
+	// Store the names only for faster access
+
+	nameIndex = FindDimLabel(procTextGrid, UTF_COLUMN, "NAME")
+	size = UTF_Utils_Vector#GetLength(procTextGrid)
+	Make/FREE=1/T/N=(size) names = procTextGrid[p][nameIndex]
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:306]: Manually copying is rather inefficient here. How about a
```igorpro
Duplicate/FREE/RMD=[0, size - 1][nameIndex] procTextGrid, names
Redimension/N=(-1) names

```

File: procedures/unit-testing-tracing.ipf [hunk 16]
```diff
@@ -960,4 +1004,46 @@ static Function AnalyzeTracingResult()
...
+		UTF_Reporting#UTF_PrintStatusMessage("Nothing to restore")
+		return NaN
+	endif
+
+	sprintf msg, "%d procedure files to restore", size
+	UTF_Reporting#UTF_PrintStatusMessage(msg)
+
+	for(i = 0; i < size; i += 1)
+		path = wvProcs[i][%PATH]
+		backupPath = path + ".backup"
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:1024]: Since the `.backup` suffix is used twice now, it should go to a `StrConstant`

File: procedures/unit-testing-tracing.ipf [hunk 17]
```diff
@@ -210,6 +225,10 @@ static Function SetupTraceProcedures(string procWinList, string traceOptions)
 		procWin = StringFromList(i, procWinList)
 		[procText, funcPath, markLines] = AddTraceFunctions(procWin, i)
 		if(!UTF_Utils#IsEmpty(funcPath))
+			gridIndex = UTF_Utils_Vector#AddRow(procTextGrid)
```
Comment (MichaelHuth) [inline procedures/unit-testing-tracing.ipf:228]: I think this breaks it, because originally procWinList is saved, here you save only a subset.

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
Currently this PR breaks tracing. I get a
`Uncaught runtime error 1321:"a wave write;Index out of range for wave "_free_"."` in every TC and UserHook.

Please check against e.g. the tests/Basic/Basic.pxp from MIES. You can start a traced test with
`RunWithOpts(instru=1)`.

File: procedures/unit-testing-tracing.ipf (hunk 14 above)
Comment (Garados007) [reply procedures/unit-testing-tracing.ipf]: Fixed it.

**Comment by Garados007** [comment]
Fixed the bugs and MIES Basic test finished successfully.

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
Now it works with the MIES test, but 
In the current approach you use the procedure names for the Z_ function for ProcTextGrid and use the same list to restore the backup files. However, it is effectively not the same list and thus, the round trip fails. For a fully correct setup I get the `Backup file not found` error for files that contain no functions or macros, because you try to restore them while they were not backuped in the first place.

**Comment by Garados007** [comment]
This will no longer prompt a `Backup file not found` for procedure files that was never instrumented or backuped because of missing functions or macros.

**Review by MichaelHuth — APPROVED** [review]
Looks good to me.
